31 write datapackagejson

iuljan opened the pull request.

FelixMau committed:
Adding Timeseries function

FelixMau committed:
Calling ad busses to elements in build datapackage

FelixMau committed:
Adding bus to parametrized elements iterative

Adding busses by calling get_bus from Mapper. Also removed `cls` parameter from get_default and get bus since the class is stored in Mapper as `adapter` anyway.

FelixMau committed:
Cleaning

FelixMau committed:
Remove old function

FelixMau committed:
Making notes

FelixMau committed:
Learned about walrus (thx)

FelixMau committed:
Conceptual Notes

FelixMau committed:
Bus for each row.

FelixMau committed:
Merge pull request #35 from sedos-project/add_bus_elements

Adding bus to parametrized elements iterative

FelixMau committed:
Merge branch '31-write-datapackagejson' into add_get_foreign_keys

# Conflicts:
#	data_adapter_oemof/build_datapackage.py

FelixMau committed:
Adding Foreign keys iterative

Adding Foreign keys while iterating for each row in scalar Data and Parametrizing.
Moved Mapper creation outside of `parametrize_dataclass` to avoid recreation of same class and to be able to make use of the class outside of `parametrize_dataclass`

Foreign keys could also be derived from the datapackage after it is successfully `build` To do that it might be necessary to make `mapper.is_sequence()` callable from outside mapper (maybe as "regular function"?)

FelixMau committed:
Removed Bug

FelixMau committed:
Improve Profile decision-making

FelixMau committed:
Using `process_name` instead of `facade_adapter_name`

FelixMau committed:
Using `process_name` instead of `facade_adapter_name`

FelixMau committed:
Saving Datapackage

Adjusted Naming of foreign_keys to be the same as in tabular, as they are named "foreignKeys"
Changed get_foreign_keys, the bus names where actually the entries that would go into "bus.csv" and not the correct fields.

For Building the Package datapackage.Package is used to infer datatypes.
1. Save elements and sequences with defined Naming
2. Build Package
3. Add Foreign keys to descriptors
4. Use updated descriptors to create new Package and save that package as json.

Then Foreign keys are added and the

FelixMau committed:
Change target directory, add Return val

FelixMau committed:
pathing

FelixMau committed:
Unify doc string format with data_adapter

Adding more comments and Extend doc strings.

FelixMau committed:
Adding test for build Datapackage

FelixMau committed:
Adding test for build Datapackage

FelixMau committed:
Adding test for build Datapackage

FelixMau committed:
Change foreignKeys to snake_case

FelixMau committed:
Merge pull request #36 from sedos-project/add_get_foreign_keys

Add get foreign keys

FelixMau committed:
Apply review

FelixMau committed:
Merge pull request #37 from sedos-project/save_datapackage

Save datapackage.
Will improve test once a propper goal datapackage is established.

FelixMau committed:
Apply Review & Fix Datapackage

Fixing Typo
Empty Bus csv is empty due to missing list() for pandas Series
Adding Primary Keys to Datapackage
Saving Elements without Index & setting correct Index and Indexlabel for Sequences
Adjusting File structure for Datapackage

FelixMau committed:
Adjusting test with temporary changes

FelixMau committed:
Changing to timeseries from data_adapter
Returning None value when no timeseries is found will lead to using default tabular value rather than illegal strings, that do not lead towards a foreign datasource.

FelixMau committed:
Extend functionality of `as_dict` to periodical values.

Extending functionality in `get_foreign_keys with more checks to narrow down on only foreign keys.

FelixMau committed:
Pre-commit

iuljan commented:
I would really like to close this PR soon. Can you give a brief overview what's still missing?

iuljan commented on data_adapter_oemof/adapters.py:
```diff
@@ -26,27 +27,40 @@ def as_dict(self):
+            if isinstance(value, oemof.solph._plumbing._Sequence):
+                # if value.periodic_values:
+                #     data[field.name] = value.periodic_values
+                if len(value) != 0:
+                    data[field.name] = value.data
```
This is not fail safe. If a "regular" sequence gets called once, it's length will be > 0. It's better to check if `periodic_values` is true.

FelixMau replied:
Yes I am aiming to use the implementation that is commented above but since I did not switch to the "periodic-sequence-plumbing" yet since it did not seem settled to me.

FelixMau replied:
I do not see a Problem in saving the "real" sequence as we would have to save it somewhere and while moving it to timeseries data would be more percise the scalarss use is not inteded for such cases as far as I know.

FelixMau commented:
I want to integrate the `_Sequence` behaiviour to correctly save the Datapackage. 
Moreover I want to write better test but this might also happen in a nother PR then.

Edit: If we are going to integrate the periodical sequence approach we also have to integrate this on in  datapackage builder as well.

FelixMau committed:
Merge branch 'dev' into 31-write-datapackagejson

FelixMau committed:
Writing datapackage with sequence check

Note: Solph branch containing updated plumbing with possible periodic values necessary!

FelixMau committed:
Merge remote-tracking branch 'origin/31-write-datapackagejson' into 31-write-datapackagejson

iuljan commented on data_adapter_oemof/adapters.py:
```diff
@@ -26,32 +27,45 @@ def as_dict(self):
+            if isinstance(value, oemof.solph._plumbing._Sequence):
+                if value.periodic_values:
+                    data[field.name] = value.periodic_values
+                elif len(value) != 0:
```
I would rather do: 
```
elif value.default: 
  data[field.name] = value.default
else:
   data[field] = value[data]

iuljan commented on tests/test_mapping.py:
`test_get_busses`, `test_default_bus_mapping`, `test_get_matched_busses` are failing:

`TypeError: Mapper.get_busses() got an unexpected keyword argument 'cls'`

iuljan commented on tests/test_build_datapackage.py:
`test_build_datapackage`, `test_build_tabular_datapackage_from_adapter` and `test_read_datapackage` fail

iuljan commented on data_adapter_oemof/build_datapackage.py:
```diff
@@ -91,56 +97,193 @@ def __split_timeseries_into_years(parametrized_sequences):
+                index_label="timeindex",
+                date_format="%Y-%m-%dT%H:%M:%SZ",
+            )
+
```
We still miss to write the information about the multi-period into  `periods.csv`. We will do this in a new PR

iuljan reviewed (commented):
Somehow poetry is not resolving, so I cant fully check the PR yet. Please check the failing tests

iuljan commented on poetry.lock:
Did you delete this by purpose?

FelixMau committed:
Merge branch 'dev' into 31-write-datapackagejson

# Conflicts:
#	poetry.lock
#	pyproject.toml
